File: library/src/main/java/com/yy/codex/uikit/UIViewTouchHandler.java
package com.yy.codex.uikit;

import android.support.annotation.NonNull;
import android.view.MotionEvent;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.Set;

/**
 * Created by saiakirahui on 2017/1/14.
 */

public class UIViewTouchHandler {

    WeakReference<UIView> viewWeakReference = null;
    UIView mHitTestedView;
    long mEventID;
    double[] mHash;

    UIViewTouchHandler(UIView view) {
        viewWeakReference = new WeakReference<UIView>(view);
    }

    void onTouchEvent(@NonNull MotionEvent event) {
        UIView view = viewWeakReference.get();
        if (view == null) {
            return;
        }
        if (event.getAction() == MotionEvent.ACTION_DOWN) {
            CGPoint touchPoint = new CGPoint(event.getX() / UIScreen.mainScreen.scale(), event.getY() / UIScreen.mainScreen.scale());
            mHitTestedView = view.hitTest(touchPoint, event);
            mEventID = System.currentTimeMillis();
        }
        if (mHitTestedView != null) {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    mHash = null;
                    mHitTestedView.touchesBegan(requestTouches(event), new UIEvent());
                    break;
                case MotionEvent.ACTION_UP:
                    mHitTestedView.touchesEnded(requestTouches(event), new UIEvent());
                    break;
                case MotionEvent.ACTION_POINTER_UP:
                case MotionEvent.ACTION_MOVE:
                case MotionEvent.ACTION_POINTER_DOWN:
                    double[] cHash = requestHash(event);
                    if (mHash != null && isHashSame(mHash, cHash)) {
                        break;
                    }
                    mHash = cHash;
                    mHitTestedView.touchesMoved(requestTouches(event), new UIEvent());
                    break;
            }
        }
    }

    UITouch[] requestTouches(MotionEvent event) {
        UIView view = viewWeakReference.get();
        if (view == null) {
            return new UITouch[0];
        }
        int pointerCount = event.getPointerCount();
        UITouch[] touches = new UITouch[pointerCount];
        double offsetX = (event.getRawX() - event.getX(0)) / UIScreen.mainScreen.scale();
        double offsetY = (event.getRawY() - event.getY(0)) / UIScreen.mainScreen.scale();
        for (int i = 0; i < pointerCount; i++) {
            double x = event.getX(i) / UIScreen.mainScreen.scale();
            double y = event.getY(i) / UIScreen.mainScreen.scale();
            CGPoint convertedPoint = view.convertPoint(new CGPoint(x, y), mHitTestedView);
            CGPoint rawPoint = new CGPoint(x + offsetX, y + offsetY);
            touches[i] = new UITouch(mHitTestedView, convertedPoint, rawPoint, requestPhase(event), mEventID);
        }
        return touches;
    }

    UITouch.Phase requestPhase(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                return UITouch.Phase.Began;
            case MotionEvent.ACTION_UP:
                return UITouch.Phase.Ended;
            case MotionEvent.ACTION_POINTER_UP:
                return UITouch.Phase.Ended;
            case MotionEvent.ACTION_MOVE:
                return UITouch.Phase.Moved;
            case MotionEvent.ACTION_POINTER_DOWN:
                return UITouch.Phase.Began;
            default:
                return UITouch.Phase.Ended;
        }
    }

    double[] requestHash(MotionEvent event) {
        int count = event.getPointerCount();
        double[] hash = new double[count * 2];
        for (int i = 0; i < count; i++) {
            hash[i] = (double)event.getX(i);
            hash[i + 1] = (double)event.getY(i);
        }
        return hash;
    }

    boolean isHashSame(double[] hashA, double[] hashB) {
        if (hashA.length == hashB.length) {
            for (int i = 0; i < hashA.length; i++) {
                if (Double.compare(hashA[i], hashB[i]) != 0) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIViewTouchHandler.kt
package com.yy.codex.uikit

import android.view.MotionEvent

import java.lang.ref.WeakReference

/**
 * Created by saiakirahui on 2017/1/14.
 */

class UIViewTouchHandler internal constructor(private val view: UIView) {

    private var hitTestedView: UIView? = null
    private var eventID: Long = 0
    private var hash: DoubleArray? = null

    internal fun onTouchEvent(event: MotionEvent) {
        if (event.action == MotionEvent.ACTION_DOWN) {
            val touchPoint = CGPoint(event.x / UIScreen.mainScreen.scale(), event.y / UIScreen.mainScreen.scale())
            hitTestedView = view.hitTest(touchPoint, event)
            eventID = System.currentTimeMillis()
        }
        hitTestedView?.let {
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    hash = null
                    it.touchesBegan(requestTouches(event), UIEvent(UIEvent.Type.Touches, UIEvent.SubType.Unknown))
                }
                MotionEvent.ACTION_UP -> it.touchesEnded(requestTouches(event), UIEvent(UIEvent.Type.Touches, UIEvent.SubType.Unknown))
                MotionEvent.ACTION_POINTER_UP, MotionEvent.ACTION_MOVE, MotionEvent.ACTION_POINTER_DOWN -> {
                    val cHash = requestHash(event)
                    if (hash is DoubleArray) {
                        if (isHashSame(hash as DoubleArray, cHash)) {
                            return
                        }
                    }
                    hash = cHash
                    it.touchesMoved(requestTouches(event), UIEvent(UIEvent.Type.Touches, UIEvent.SubType.Unknown))
                }
            }
        }
    }

    internal fun requestTouches(event: MotionEvent): List<UITouch> {
        val hitTestedView = hitTestedView ?: return emptyList()
        val pointerCount = event.pointerCount
        val touches: MutableList<UITouch> = mutableListOf()
        val offsetX = (event.rawX - event.getX(0)) / UIScreen.mainScreen.scale()
        val offsetY = (event.rawY - event.getY(0)) / UIScreen.mainScreen.scale()
        for (i in 0..pointerCount - 1) {
            val x = event.getX(i) / UIScreen.mainScreen.scale()
            val y = event.getY(i) / UIScreen.mainScreen.scale()
            val convertedPoint = view.convertPoint(CGPoint(x, y), hitTestedView)
            val rawPoint = CGPoint(x + offsetX, y + offsetY)
            touches.add(UITouch(hitTestedView, convertedPoint, rawPoint, requestPhase(event), eventID))
        }
        return touches.toList()
    }

    internal fun requestPhase(event: MotionEvent): UITouch.Phase {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> return UITouch.Phase.Began
            MotionEvent.ACTION_UP -> return UITouch.Phase.Ended
            MotionEvent.ACTION_POINTER_UP -> return UITouch.Phase.Ended
            MotionEvent.ACTION_MOVE -> return UITouch.Phase.Moved
            MotionEvent.ACTION_POINTER_DOWN -> return UITouch.Phase.Began
            else -> return UITouch.Phase.Ended
        }
    }

    internal fun requestHash(event: MotionEvent): DoubleArray {
        val count = event.pointerCount
        val hash = DoubleArray(count * 2)
        for (i in 0..count - 1) {
            hash[i] = event.getX(i).toDouble()
            hash[i + 1] = event.getY(i).toDouble()
        }
        return hash
    }

    internal fun isHashSame(hashA: DoubleArray, hashB: DoubleArray): Boolean {
        if (hashA.size == hashB.size) {
            return hashA.indices.none { java.lang.Double.compare(hashA[it], hashB[it]) != 0 }
        }
        return false
    }

}
